src/hotspot/share/classfile/fieldLayoutBuilder.cpp

```diff
@@ -658,7 +658,7 @@ void FieldLayoutBuilder::regular_field_sorting() {
             //too_volatile_to_flatten = false; //FIXME
             // volatile fields are currently never flatten, this could change in the future
           }
-          if (!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten) || fieldinfo.access_flags().is_final()) {
```

Comment:
An IR framework test can be added to check for flattening of Primitive class final/non-final fields.
Given that InlineTypeNode are removed during Optimization check has to done on post Parse IR.